src/moaicore/MOAICCParticleSystem.cpp

```diff
@@ -0,0 +1,2385 @@
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_startSystem ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StartSystem();
+    return 0;
+}
+
+//----------------------------------------------------------------//
+/** @name   stopSystem
+    @text   Halt the generation of particles.  Any remaining particles will be allowed to stay until the end of their lifespan. Use stop() to halt the animation of particles.
+ 
+    @in     MOAICCParticleSystem    self
+    @out    nil
+ */
+int MOAICCParticleSystem::_stopSystem ( lua_State *L ) {
+    
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    self->StopSystem();
+    
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   surge
+    @text   Generate the given number of particles at one time.
+    @in     MOAICCParticleSystem    self
+    @in     number                  numberOfParticles
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_surge ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "UN" )
+    u32 numberOfParticles = state.GetValue < u32 >(2, 1.0);
+    self->Surge( numberOfParticles );
+    return 0;
+}
+//----------------------------------------------------------------//
+/** @name   reset
+    @text   Make all particles get cleared out at the next frame.
+ 
+    @in     MOAICCParticleSystem    self
+    @opt    boolean                 activate    Default true.  Whether to begin the generation of particles after clearing the particles.
+    @out    nil
+ */
+
+int MOAICCParticleSystem::_reset ( lua_State *L ) {
+    MOAI_LUA_SETUP( MOAICCParticleSystem, "U" )
+    bool activate = state.GetValue < bool > (2, true);
+    self->ResetSystem( activate );
+    return 0;
+}
+
+//----------------------------------------------------------------//
+//----------------------------------------------------------------//
+
+bool MOAICCParticleSystem::AddParticle () {
+    if ( this->IsFull() ) {
+        return false;
+    }
+    
+	MOAICCParticle * particle = new MOAICCParticle();
+    this->InitParticle( particle );
+	this->mParticles.push_back(*particle);
+	
+    this->mParticleCount++;
+    
+    return true;
+}
+
+void MOAICCParticleSystem::Draw ( int subPrimID ) {
+    UNUSED(subPrimID);
+    
+    if ( !this->mDeck ) {
+        return;
+    }
+    
+    MOAIGfxDevice& gfxDevice = MOAIGfxDevice::Get ();
+    
+    if ( this->mUVTransform ) {
+        USAffine3D uvMtx = this->mUVTransform->GetLocalToWorldMtx ();
+        gfxDevice.SetUVTransform ( uvMtx );
+    }
+    else {
+        gfxDevice.SetUVTransform ();
+    }
+    
+    this->LoadGfxState();
+    
+    USAffine3D drawingMtx;
+    USAffine3D spriteMtx;
+    
+    std::vector<MOAICCParticle>::iterator end = this->mParticles.end();
+    for (std::vector<MOAICCParticle>::iterator particleIterator = this->mParticles.begin() ; particleIterator != end; ++particleIterator) {
```

Comment:
Are we using a C++11 compiler? If so, you can do this:

    for (auto particle: this->mParticles)